### ISSUE TITLE
Nested attributes issue with POST request.

### ISSUE THREAD

**Issue by gauravgupta121** [visualitypl/jsonapi_parameters#39]
Hello,

I'm trying `jsonapi_parameters`gem for structuring the request parameters for a has-one relationship with nested attributes.

ruby '2.6.3'
rails  '6.0.1'

Below is my code:

```ruby

#Company Model
class Company < ApplicationRecord
  has_one :company_address

  accepts_nested_attributes_for :company_address
end

#Company Address Model
class CompanyAddress < ApplicationRecord
  belongs_to :company
end

#Companies Controller
class Api::V2::CompaniesController < ApplicationController

  def create
    company = Company.new(company_params)
  end

  def company_params
    params.from_jsonapi.require(:company).permit(
      :business_name,
      :business_email,
      :mobile_number,
      :business_pan,
      :number_of_directors,
      :business_type,
      :date_inc,
      :gst_number,
      company_address_attributes: [
        :address_line1,
        :address_line2, 
        :city,
        :state,
        :pincode,
        :address_type,
        :ownership_type
      ]
    )
  end
end
```

**CASE-1**

Request:

URL: http://localhost:4000/v2/companies
Method: POST

params:

```ruby
{
    "data": {
        "type": "company",
        "attributes": {
            "business_name": "New Looks Garments",
            "business_email": "abc@example.com",
            "mobile_number": "1234567890",
            "business_pan": "ABCDE7412A",
            "number_of_directors": 1,
            "business_type": "proprietorship",
            "date_inc": "2014-07-19",
            "gst_number": "27ABCDE660052Z0"
        },
        "relationships": {
            "company_address": {
                "data": {
                    "type": "company_address"
                }
            }
        }
    },
    "included": [
        {
            "type": "company_address",
            "attributes": {
                "address_line1": "marcant road",
                "address_line2": "address 21",
                "city": "JALPAIGURI",
                "state": "West Bengal",
                "pincode": "226201",
                "address_type": "residence",
                "ownership_type": "owned"
            }
        }
    ]
}
```

Response:

value of company_params:

```ruby
Unpermitted parameter: :company_address_id
{
          "business_name" => "New Looks Garments",
         "business_email" => "abc@example.com",
          "mobile_number" => "1234567890",
           "business_pan" => "ABCDE7412A",
    "number_of_directors" => 1,
          "business_type" => "proprietorship",
               "date_inc" => "2014-07-19",
             "gst_number" => "27ABCDE660052Z0"
}
```

Here it is not including the relationships in the formed request parametes.

**CASE-2**

URL: http://localhost:4000/v2/companies
Method: POST

Now when I'm passing the `id` in the `relationships` and `included` nodes then it is returning in the correct form.
But ideally `id` should not be required in case of new-record.

**Request:**

```ruby
{
    "data": {
        "type": "company",
        "attributes": {
            "business_name": "New Looks Garments",
            "business_email": "gauravv@emaill.com",
            "mobile_number": "8188006587",
            "business_pan": "TTRPS7412A",
            "number_of_directors": 1,
            "business_type": "proprietorship",
            "date_inc": "2014-07-19",
            "gst_number": "27AMAPY660052Z0"
        },
        "relationships": {
            "company_address": {
                "data": {
                    "id": 1,
                    "type": "company_address"
                }
            }
        }
    },
    "included": [
        {
            "id": 1,
            "type": "company_address",
            "attributes": {
                "address_line1": "marcant road",
                "address_line2": "address 21",
                "city": "JALPAIGURI",
                "state": "West Bengal",
                "pincode": "226201",
                "address_type": "residence",
                "ownership_type": "owned"
            }
        }
    ]
}
```
Response of `company_params`:

```ruby
Unpermitted parameter: :id
{
                 "business_name" => "New Looks Garments",
                "business_email" => "abc@example.com",
                 "mobile_number" => "1234567890",
                  "business_pan" => "ABCDE7412A",
           "number_of_directors" => 1,
                 "business_type" => "proprietorship",
                      "date_inc" => "2014-07-19",
                    "gst_number" => "27ABCDE660052Z0",
    "company_address_attributes" => {
         "address_line1" => "abc road",
         "address_line2" => "address 21",
                  "city" => "JALPAIGURI",
                 "state" => "West Bengal",
               "pincode" => "226201",
          "address_type" => "residence",
        "ownership_type" => "owned"
    }
}

```

So now I'm not able to create the new nested objects in POST request without passing the `id` in the `relationships` and `included` nodes.

I'm eager to receive our feedback.

Thanks!

**Comment by Marahin** [comment]
Hello,

thanks for reaching out and providing detailed steps on how to reproduce the issue. We'll investigate and get back to you shortly!

**Comment by gauravgupta121** [comment]
Hi @Marahin ,

Thanks for the quick response.

**Comment by Marahin** [comment]
Hello,
so after initial investigation it seems that your parameters are not JSON:API compliant.
According to the documentation at [jsonapi.org, in section of Compound Documents](https://jsonapi.org/format/#document-compound-documents):
>To reduce the number of HTTP requests, servers MAY allow responses that include related resources along with the requested primary resources. Such responses are called “compound documents”.
>In a compound document, all included resources MUST be represented as an array of resource objects in a top-level `included` member.
Indeed, [Resource Objects](https://jsonapi.org/format/#document-resource-objects) have an exception of not providing the `id` in case its a new record, but that is not the case here. If you continue reading the Compound Documents section, the next paragraph states:
 >Compound documents require “full linkage”, meaning that every included resource MUST be identified by at least one resource identifier object in the same document. These resource identifier objects could either be primary data or represent resource linkage contained within primary or included resources.
So, in other words, when using compound documents to create, edit or basically use JSON API with Rails nested attributes and included tree, you have to ensure linkage is in tact, e.g. the relationships are directly pointing to a single object in included tree, and for that you need to ensure both of these objects have `id` and `type` specified. Otherwise it would be unknown what resource belongs to which entity.
You could use some UUID-generation libraries and use these as your temporary, client-side IDs used only for linking entities to one another.
